packages/perseus/src/widgets/python-program.tsx

```diff
@@ -0,0 +1,83 @@
+/**
+ * This widget is for embedding Khan Academy Python programs.
+ */
+
+import {StyleSheet, css} from "aphrodite";
+import PropTypes from "prop-types";
+import * as React from "react";
+import _ from "underscore";
+
+import {getDependencies} from "../dependencies";
+import {isFileProtocol} from "../util/mobile-native-utils";
+import {toAbsoluteUrl} from "../util/url-utils";
+
+import type {WidgetExports} from "../types";
+
+function getUrlFromProgramID(programID: any) {
+    const {InitialRequestUrl} = getDependencies();
+
+    const path = `/python-program/${programID}/embedded`;
+
+    return toAbsoluteUrl(path);
+}
+
+/* This renders the program in an iframe. */
+class PythonProgram extends React.Component<any> {
+    static propTypes = {
+        programID: PropTypes.string,
+        height: PropTypes.number,
+    };
+
+    static defaultProps: any = {
+        height: 400,
+    };
+
+    render(): React.ReactNode {
+        if (!this.props.programID) {
```

Comment:
Is there a reason why we'd ever allow a null/undefined `programID`? If not, then the editor should disallow it and we should type the `PerseusPythonProgramWidgetOptions` as having a required `programID` field.

Reply:
I'm not completely sure how to disallow an empty `programID` in the editor. I added a `getSaveWarnings` function, but I don't know how to test it.

Reply:
It looks like you resolved this. The `PerseusPythonProgramWidgetOptions` should mark the field(s) as non-optional and then in the widget you should be able to depend on the types optionality.

Reply:
I also just figured out that a widget editor can implement a function `getSaveWarnings` which is basically a validation function for saving the widget. You can implement save validation there to ensure that there is a Program ID and valid Height in that function.

Reply:
I've added validation of both options to `getSaveWarnings` and added tests for the validation logic.